### PR TITLE
server: small refactor for grpc-server

### LINKED ISSUE #4675: grpc-service has many similar code to handler the forward request.
## Enhancement Task
some similar code can be exracted, like this :
https://github.com/tikv/pd/blob/4b6f00c3ed0ee52d7cc60b15a446357a7d27b53b/server/grpc_service.go#L388-L400

### PR DESCRIPTION
**Pull request by bufferflies**
Signed-off-by: bufferflies <1045931706@qq.com>

### What problem does this PR solve?

Issue Number: Close #4675

### What is changed and how it works?

```commit-message
```

### Check List

Tests

- Unit test

Code changes

Side effects

Related changes

### Release note

```release-note
None
```

### CODE REVIEW

File: server/grpc_service.go [hunk 1]
```diff
@@ -59,6 +59,28 @@ var (
 	ErrNotStarted = status.Errorf(codes.Unavailable, "server not started")
 )
 
+type Header interface {
```
Comment (JmPotato) [inline server/grpc_service.go:62]: Comments are needed.

**Review by JmPotato — COMMENTED** [review]
The reset LGTM.

File: server/grpc_service.go [hunk 2]
```diff
@@ -59,6 +59,28 @@ var (
...
+	forwardedHost := getForwardedHost(ctx)
+	if !s.isLocalRequest(forwardedHost) {
+		client, err := s.getDelegateClient(ctx, forwardedHost)
+		if err != nil {
+			return nil, err
+		}
+		ctx = grpcutil.ResetForwardContext(ctx)
+		return fn(ctx, client)
+	}
+	if err := s.validateRequest(header.GetHeader()); err != nil {
```
Comment (rleungx) [inline server/grpc_service.go:78]: I prefer not to contain it in `forwardRequest`.

**Review by disksing — APPROVED** [review]

File: server/grpc_service.go (hunk 1 above)
Comment (bufferflies) [reply server/grpc_service.go:62]: fix

File: server/grpc_service.go (hunk 2 above)
Comment (bufferflies) [reply server/grpc_service.go:79]: I think the forward request should validate the request also .

File: server/grpc_service.go [hunk 3]
```diff
@@ -59,6 +59,29 @@ var (
 	ErrNotStarted = status.Errorf(codes.Unavailable, "server not started")
 )
 
+// Header returns header metadata.
+type Header interface {
+	GetHeader() *pdpb.RequestHeader
```
Comment (rleungx) [inline server/grpc_service.go:64]: We don't need this interface. How about using `ResponseHeader` directly?

File: server/grpc_service.go (hunk 2 above)
Comment (rleungx) [reply server/grpc_service.go:79]: Yes, but the function name should be changed. For now, it sounds like only forward requests need to be validated.

File: server/grpc_service.go (hunk 3 above)
Comment (bufferflies) [reply server/grpc_service.go:65]: fix

File: tests/server/global_config/global_config_test.go [hunk 4]
```diff
@@ -235,11 +235,3 @@ func (s *GlobalConfigTestSuite) TestClientWatch(c *C) {
 		}
 	}
 }
-
-func (s *GlobalConfigTestSuite) TestClientWatchTimeout(c *C) {
```
Comment (JmPotato) [inline tests/server/global_config/global_config_test.go:239]: Could we delete this gRPC method directly?

Comment (bufferflies) [reply tests/server/global_config/global_config_test.go:239]: the test function is not steady, and  the global config is not using.

**Review by rleungx — APPROVED** [review]

**Comment by JmPotato** [comment]
CI failed. PTAL @bufferflies 

/merge cancel

File: server/grpc_service.go [hunk 5]
```diff
@@ -350,15 +371,12 @@ func watchTSDeadline(ctx context.Context, tsDeadlineCh <-chan deadline) {
...
-	if !s.isLocalRequest(forwardedHost) {
-		client, err := s.getDelegateClient(ctx, forwardedHost)
-		if err != nil {
-			return nil, err
-		}
-		ctx = grpcutil.ResetForwardContext(ctx)
+	fn := func(ctx context.Context, client *grpc.ClientConn) (interface{}, error) {
 		return pdpb.NewPDClient(client).Bootstrap(ctx, request)
 	}
+	if rsp, err := s.unaryMiddleware(ctx, request.GetHeader(), fn); err == nil && rsp != nil {
```
Comment (rleungx) [inline server/grpc_service.go:377]: We should handle the cases when the error is not nil.

Comment (bufferflies) [reply server/grpc_service.go:377]: yes , it will return err if err is not nil

**Comment by bufferflies** [comment]
PTAL @rleungx  again

**Review by rleungx — APPROVED** [review]

File: server/grpc_service.go [hunk 6]
```diff
@@ -350,15 +371,14 @@ func watchTSDeadline(ctx context.Context, tsDeadlineCh <-chan deadline) {
...
+	fn := func(ctx context.Context, client *grpc.ClientConn) (interface{}, error) {
 		return pdpb.NewPDClient(client).Bootstrap(ctx, request)
 	}
+	if rsp, err := s.unaryMiddleware(ctx, request.GetHeader(), fn); err != nil {
+		return nil, err
+	} else if rsp != nil {
+		return rsp.(*pdpb.BootstrapResponse), nil
+	}
 
 	if err := s.validateRequest(request.GetHeader()); err != nil {
```
Comment (rleungx) [inline server/grpc_service.go:383]: remove it?

Comment (bufferflies) [reply server/grpc_service.go]: fix

**Pull request merged**